pkg/pgmodel/cache/series_cache.go

```diff
@@ -164,48 +84,59 @@ func (t *SeriesCacheImpl) GetSeriesFromLabels(ls labels.Labels) (*model.Series,
 		ll[i].Name = ls[i].Name
 		ll[i].Value = ls[i].Value
 	}
-	l, _, err := t.GetSeriesFromProtos(ll)
+	l, err := t.GetSeriesFromProtos(ll)
 	return l, err
 }
 
-// useByteAsStringNoCopy uses a byte slice as a string in a function callback
-// without allocation of a copy. This is potentially unsafe and should only
-// be used for performance critical code. The string passed to the function is
-// only valid during the function call. Thus, you should make sure the callback
-// function does not save the string through closures or inside other datastructures.
-// the callback should not modify the byte slice.
-//
-// This usage convention guarantees the byte slice is valid during the callback
-// call but makes no other guarantees.
-func useByteAsStringNoCopy(b []byte, useAsStringFunc func(string)) {
-	//use a byte to String conversion directly without allocation
-	str := *(*string)(unsafe.Pointer(&b)) // #nosec
-	useAsStringFunc(str)
-}
-
 // GetSeriesFromProtos returns a model.Series entry given a list of Prometheus
 // prompb.Label.
 // If the desired entry is not in the cache, a "placeholder" model.Series entry
 // is constructed and put into the cache. It is not populated with database IDs
 // until a later phase, see model.Series.SetSeriesID.
-func (t *SeriesCacheImpl) GetSeriesFromProtos(labelPairs []prompb.Label) (*model.Series, string, error) {
-	builder := keyPool.Get().(*bytes.Buffer)
-	builder.Reset()
-	defer keyPool.Put(builder)
-	metricName, err := generateKey(labelPairs, builder)
-	if err != nil {
-		return nil, "", err
-	}
-	var series *model.Series
-	useByteAsStringNoCopy(builder.Bytes(), func(key string) {
-		series = t.loadSeries(key)
-	})
+func (t *SeriesCacheImpl) GetSeriesFromProtos(labelPairs []prompb.Label) (*model.Series, error) {
+	seriesKey := t.getSeriesKey(labelPairs)
+	series := t.loadSeries(seriesKey)
 	if series == nil {
-		//this will allocate the key
-		key := builder.String()
-		series = model.NewSeries(key, labelPairs)
-		series = t.setSeries(key, series)
+		series = model.NewSeries(labelPairs)
+		series = t.setSeries(seriesKey, series)
 	}
+	return series, nil
+}
 
-	return series, metricName, nil
+// We use dictionary encoding to generate series key.
+// Each labelPair is encoded to uint32
+// We perform variable encoding on uint32 to byte and append all to byte buffer
+// Bytes are then serialized to string using Base64 encoding
+func (s *SeriesCacheImpl) getSeriesKey(labelPairs []prompb.Label) string {
+	seriesKeyBytes := keyPool.Get().(*bytes.Buffer)
+	idEncodeBuf := make([]byte, binary.MaxVarintLen32)
+	ensureLabelsSorted(labelPairs)
+	for _, labelPair := range labelPairs {
+		l := labels.Label{Name: labelPair.Name, Value: labelPair.Value}
```

Comment:
I think it would be much efficient to use prompb.Label as is(may be prompb.Label.Name + prompb.Label.Value) instead of converting to labels.Label which is a map?